### ISSUE TITLE
can you support flex-wrap and hug content( without fixed width and height)

### ISSUE THREAD

**Issue by kaneruan** [bernaferrari/FigmaToCode#82]
can you support flex-wrap and hug content( without fixed width and height)

**Comment by bernaferrari** [comment]
Yes, I just need to figure out the new API.

**Comment by davestewart** [comment]
Hey, have the same issue here.

Seems like [this line](https://github.com/bernaferrari/FigmaToCode/blob/main/packages/backend/src/tailwind/builderImpl/tailwindSize.ts#L17) could just be:

```ts
  if (typeof size.width === "number" && (node as FrameNode).layoutSizingHorizontal === "FIXED") {
```

![CleanShot 2024-07-02 at 17 21 27](https://github.com/bernaferrari/FigmaToCode/assets/132681/6e2b76c4-be49-4f23-a3c7-30c8ec814482)

**Comment by bernaferrari** [comment]
if it works well feel free to make a PR :P but it should be "layoutSizingHorizontal" in node && node.layoutSizingHorizontal, FrameNode is too restrictive. InstanceNode, ComponentNode, other things might have it too.

**Comment by davestewart** [comment]
Hey @bernaferrari, @kaneruan ...

I'm looking at this, this morning, and continuing the discussion her (vs the [previous PR](https://github.com/bernaferrari/FigmaToCode/pull/111)) so others can contribute.

It seems that to better-support Flexbox and Tailwind that some more work is going to need to be done.

### Flexbox / sizing

Currently the HTML and Tailwind exports do not correctly support the various fixed/fill combinations.

Here, the boxes have fixed widths:

![CleanShot 2024-07-03 at 10 39 14](https://github.com/bernaferrari/FigmaToCode/assets/132681/b755e531-2097-4c4d-9bca-7af06989ac03)

Yet, they export with flexbox as scaled to fit:

![CleanShot 2024-07-03 at 10 35 05](https://github.com/bernaferrari/FigmaToCode/assets/132681/8993918e-a9ff-41ae-bfa1-3192c1ea53b9)

They should only do this if set to fill container:

![CleanShot 2024-07-03 at 10 39 28](https://github.com/bernaferrari/FigmaToCode/assets/132681/c5c9d8ef-bd33-4833-bc3e-ac3c7eee5895)

Setting `min-width` would solve this:

![CleanShot 2024-07-03 at 10 30 47](https://github.com/bernaferrari/FigmaToCode/assets/132681/a03d764a-e7af-4860-8656-a2541c2cd53c)

And would also work in Tailwind:

![CleanShot 2024-07-03 at 10 36 58](https://github.com/bernaferrari/FigmaToCode/assets/132681/a1feece6-e045-4726-8108-58b2ecc4688f)

### Flexbox / axis

To achieve wrapping, I'll need to tweak the `align-items` / `align-content` CSS properties when wrapping is on, but I seem to have this working:

![CleanShot 2024-07-03 at 10 54 48](https://github.com/bernaferrari/FigmaToCode/assets/132681/6bb9f8b7-9fc6-4c66-b080-9f344b1237e0)

### In summary

If everyone is happy with this, I'll implement the following:

- [ ] `min-width` / `min-height` for fixed-width content
- [x] `align-items` / `align-content` for wrapped content

LMK, thanks.

**Comment by davestewart** [comment]
@bernaferrari – what is the `optimizeLayout` flag supposed to do?

It seems to just prefer `inferredAutoLayout.*` to `node.*` in the functions it's used it.

I see this is a user checkbox, vs something used internally.

I understand that [the property](https://www.figma.com/plugin-docs/api/FrameNode/#inferredautolayout) is a kind of amalgamation of node properties. I've run some test code and it appears that the two are always in sync. 

Does it make a difference to the code output?

Am I missing the intent?

**Comment by bernaferrari** [comment]
optimize layout is figma trying to infer the AutoLayout for you. If you
have a frame with two rectangles, it is going to pretend there is an
AutoLayout. So the user can be "I prefer the system to figure out
AutoLayout for me" or "I'll do myself". With the new "magic AutoLayout"
system they introduced last week, it will be super easy for people to make
AutoLayout everywhere, so I guess inferred auto layout will be less needed.

**Comment by bernaferrari** [comment]
inferred doesn't return the childrens ordered, so I try to guess their order :P

**Comment by davestewart** [comment]
Sure, but when an object already has auto-layout, ordering does not need to be guessed.

My point is, this function is being called from `<platform>Frame()` even when auto-layout is on (and I don't think it should be).

![CleanShot 2024-07-17 at 00 40 36](https://github.com/user-attachments/assets/5bde3d8b-a611-48ac-bc56-32c5dbd08681)

Am I wrong?

**Comment by bernaferrari** [comment]
oh, interesting. Yeah, you found a bug!!

**Comment by davestewart** [comment]
OK! Thanks for confirming. I'll fix it along with the wrapping implementation.

**Comment by davestewart** [comment]
So I finally figured out what _inferred_ auto layout is 😅

Any layout for which you could click "Auto Layout" and the layout would be rendered **exactly the same** once auto-layout properties such as gap, padding and alignment were applied.

![CleanShot 2024-07-17 at 01 05 14](https://github.com/user-attachments/assets/27247894-0c9f-4bb4-97b0-148846247522)

Not sure why they don't say this in the docs!

**Comment by bernaferrari** [comment]
Yes. I thought I was clear, sorry.

**Comment by davestewart** [comment]
No worries. Thanks for helping get up to speed on the Figma API and answering all my questions!

**Comment by davestewart** [comment]
Interestingly, `inferredAutoLayout` is false for potentially wrapped elements, even when clicking the auto-layout button correctly applies wrap to the layout:

![CleanShot 2024-07-17 at 01 15 50](https://github.com/user-attachments/assets/1fb441ed-2e2a-4dec-9552-bb4e11d87ed0)

Suspect the heuristics only take into account the general offsets and padding, etc.

**Comment by davestewart** [comment]
FWIW, some of the options could be disabled if they are not relevant:

- Optimize Layout – if there are no `inferredAutoLayout` nodes
- Round Values, Round Colors, and Custom Colors, if no `[]` or `/` characters found in classes

**Comment by bernaferrari** [comment]
if I enable inferredAutoLayout, click in a layout without, then click in a layout with it, will the option work? 

If it is documented somewhere that they are disabled because there is nothing, I guess it is fine. Or could have some sort of section like "unused on current selection:"

**Comment by davestewart** [comment]
Ah, I see what you mean. 

I suppose I was thinking about making it clearer for users who might click the options (when irrelevant) and wonder why they were not having any effect.

Not a big deal, just a nice-to-have.

**Comment by bernaferrari** [comment]
we can have that disabled, as long as it is clear for users (even if irrelevant). Another way is add on the tooltip: "current selection doesn't make use of color variables". with a check or X. Could be useful for people.

**Comment by davestewart** [comment]
Will add a feature request issue.

**Comment by bernaferrari** [comment]
Apparently flex-wrap is hard, but hug, max-width, max-height, min-width, min-height are now supported (https://github.com/bernaferrari/FigmaToCode/pull/184)
